packages/flutter/lib/src/widgets/text_selection.dart

```diff
@@ -2012,6 +2015,28 @@ class TextSelectionGestureDetectorBuilder {
   // cursor will not move on drag update.
   bool? _dragBeganOnPreviousSelection;
 
+  /// Handler for [TextSelectionGestureDetector.onTapTrackStart].
+  ///
+  /// See also:
+  ///
+  ///  * [TextSelectionGestureDetector.onTapTrackStart], which triggers this
+  ///    callback.
+  @protected
+  void onTapTrackStart() {
+    _keysPressedOnDown = HardwareKeyboard.instance.logicalKeysPressed;
```

Comment:
directly calling `HardwareKeyboard.instance` from this code means it can't be easily overridden, ideally we'd be receiving the keyboard instance from the class that created this one, or via the binding.

Reply:
The use of `HardwareKeyboard.instance` came from the original PR that landed the new recogniser - it's similar to that in `MenuAnchor`.  Below the getter is:
```
static HardwareKeyboard get instance => ServicesBinding.instance.keyboard;
```
I'm guessing this is enough to cover all test scenarios. Wondering if @Renzo-Olivares might have more thoughts.

Reply:
This line originally came from `TextSelectionGestureDetector` as well. The original code block is below.

```
  // Returns true iff either shift key is currently down.
  bool get _isShiftPressed {
    return HardwareKeyboard.instance.logicalKeysPressed
      .any(<LogicalKeyboardKey>{
        LogicalKeyboardKey.shiftLeft,
        LogicalKeyboardKey.shiftRight,
      }.contains);
  }
```

Reply:
I don't think there are many overriders of `TextSelectionGestureDetectorBuilder` (besides the framework material and cupertino TextFields) I think mostly because it holds the default behavior for text selection in a text field. We do expose some callbacks through `TextField` but thats about it. Someone could possibly extend it though but they would also have to extend `TextField`.